### PR TITLE
Allow configurable npmLocation for typingsInstaller

### LINKED ISSUE #15287: ATA Does Not Work When Only Using NVM
**TypeScript Version:**  2.3.1-insiders0416

**Steps to reproduce**
On a machine that only has node installed using `nvm` and not globally (i.e. not in `/usr/local/bin`), trigger ATA through VSCode by adding an new package import to a javascript file

**bug**
ATA fails in this case and says that `npm` in not installed.

```
Sending response: {"kind":"event::beginInstallTypes","eventId":1,"typingsInstallerVersion":"2.3.1-insiders.20170416","projectName":"/Users/matb/projects/sand/jsconfig.json"}
Response has been sent.
#1 with arguments'["@types/ace@ts2.3"]'.
npm install #1 took: 7 ms
stdout: 
stderr: /bin/sh: npm: command not found

install request failed, marking packages as missing to prevent repeated requests: ["ace"]
Sending response: {"kind":"event::endInstallTypes","eventId":1,"projectName":"/Users/matb/projects/sand/jsconfig.json","packagesToInstall":["@types/ace@ts2.3"],"installSuccess":false,"typingsInstallerVersion":"2.3.1-insiders.20170416"}
Response has been sent.
```

In this case, `npm` can be run from the user's command prompt just fine. It is linked to the nvm location for example: `/Users/matb/.nvm/versions/node/v7.6.0/bin/npm`

### PR DESCRIPTION
**Pull request by andy-ms**
Fixes #15287

When `--npmLocation /foo/npm` is passed to the server, the server will pass the same argument to typingsInstaller, and typingsInstaller will use `/foo/npm install ...` in place of `npm install ...`.

(This does not include a unit test :frowning:, that would require launching new processes to ensure that the arguments get through. Our unit tests for typingsInstaller use a TestTypingsInstaller that avoids using NPM.)

### CODE REVIEW

File: src/server/server.ts [hunk 1]
```diff
@@ -278,19 +280,21 @@ namespace ts.server {
...
-                    if (match) {
-                        // if port is specified - use port + 1
-                        // otherwise pick a default port depending on if 'debug' or 'inspect' and use its value + 1
-                        const currentPort = match[3] !== undefined
-                            ? +match[3]
-                            : match[1] === "debug" ? 5858 : 9229;
-                        execArgv.push(`--${match[1]}=${currentPort + 1}`);
-                        break;
-                    }
+            for (const arg of process.execArgv) {
```
Comment (andy-ms) [inline src/server/server.ts]: This was needlessly put in a nested block.

File: src/server/typingsInstaller/nodeTypingsInstaller.ts [hunk 2]
```diff
@@ -86,7 +87,7 @@ namespace ts.server.typingsInstaller {
             if (this.log.isEnabled()) {
                 this.log.writeLine(`Process id: ${process.pid}`);
             }
-            this.npmPath = getNPMLocation(process.argv[0]);
+            this.npmPath = npmLocation !== undefined ? npmLocation : getDefaultNPMLocation(process.argv[0]);
```
Comment (mhegazy) [inline src/server/typingsInstaller/nodeTypingsInstaller.ts]: can you add a log entry for the location of npm we will end up using.

File: src/server/server.ts [hunk 3]
```diff
@@ -219,7 +220,7 @@ namespace ts.server {
         }
     }
 
-    class NodeTypingsInstaller implements ITypingsInstaller {
+    export class NodeTypingsInstaller implements ITypingsInstaller {
```
Comment (mhegazy) [inline src/server/server.ts]: why `export`?

Comment (andy-ms) [inline src/server/server.ts]: Whoops, was trying to undo the change I made to the other `NodeTypingsInstaller`.

File: src/server/typingsInstaller/nodeTypingsInstaller.ts [hunk 4]
```diff
@@ -84,10 +84,11 @@ namespace ts.server.typingsInstaller {
                 typingSafeListLocation ? toPath(typingSafeListLocation, "", createGetCanonicalFileName(sys.useCaseSensitiveFileNames)) : toPath("typingSafeList.json", __dirname, createGetCanonicalFileName(sys.useCaseSensitiveFileNames)),
                 throttleLimit,
                 log);
+            this.npmPath = npmLocation !== undefined ? npmLocation : getDefaultNPMLocation(process.argv[0]);
             if (this.log.isEnabled()) {
                 this.log.writeLine(`Process id: ${process.pid}`);
+                this.log.writeLine(`NPM location: ${npmLocation}`);
```
Comment (mhegazy) [inline src/server/typingsInstaller/nodeTypingsInstaller.ts]: can we log `this.npmPath` as well.

**Comment by mhegazy** [comment]
@andy-ms please port this to release-2.3 as well.

**Comment by mhegazy** [comment]
//cc @mjbvz

**Comment by mhegazy** [comment]
thanks!
